front/lib/resources/vault_resource.ts

```diff
@@ -391,107 +392,95 @@ export class VaultResource extends BaseResource<VaultModel> {
     });
   }
 
+  private mapGroupPermissions(permissions: Permission[]) {
+    return this.groups.map((group) => ({
+      id: group.id,
+      permissions,
+    }));
+  }
+
   /**
-   * Determine ACL entries based on vault type and group configuration:
-   * 1. For regular vaults with a global group:
-   *    - Return only the global group with full permissions
-   *    - Ignore regular groups as they're not used in non-restricted vaults
-   * 2. For all the other vaults:
-   *    - Return all associated groups with full permissions
+   * Computes Access Control List (ACL) entries based on vault type and group configuration.
+   *
+   * Permission patterns:
+   * - System vaults: admin-only write access
+   * - Public vaults: anyone can read, workspace members can write
+   * - Global vaults: admin and builder have read/write access
+   * - Regular vaults with global group: hierarchical access (admin > builder > user)
+   * - Restricted vaults: admin access + explicit group permissions
    */
-  acl(): ACLType {
+  acl(): GroupAndRoleACL {
     const globalGroup = this.isRegular()
       ? this.groups.find((group) => group.isGlobal())
       : undefined;
+
+    if (this.isSystem()) {
+      return {
+        workspaceId: this.workspaceId,
+        roles: [{ name: "admin", permissions: ["admin", "write"] }],
```

Comment:
Here we should not have write ideally. An admin should not be able to mess with the data. 

This likely requires changing a few things in the resource logic and possibly endpoint but would provide much higher guarantees right?

Reply:
I'm happy to do so, but it requires changing some logic both in the UI and backend as we mainly check `canWrite` when connecting managed data sources. I suggest I do this in a follow up PR so we can limit the scope of this one to the ACL logic refactoring.

Reply:
Let's create a task for that then. Real must have if you ask me 🙏